v3data/hypes/impermanent_data.py

```diff
@@ -0,0 +1,441 @@
+            },
+            limitUpper: ticks(
+                block: {number: $block}
+                where: {
+                poolAddress: $poolAddress
+                tickIdx: $limitUpper
+                }
+            ){
+                tickIdx
+                feeGrowthOutside0X128
+                feeGrowthOutside1X128
+            }
+        }
+        """
+
+        variables = {
+            "block": int(block),
+            "poolAddress": pool_address,
+            "baseLower": base_lower,
+            "baseUpper": base_upper,
+            "limitLower": limit_lower,
+            "limitUpper": limit_upper,
+        }
+
+        response = await self.uniswap_client.query(pool_query, variables)
+
+        return response["data"]
+
+    async def get_impermanent_data(self, get_data=True):
+        
+        if get_data:
+            await self.get_data()
+
+        
+        # get hypervisors data
+        hypervisor_dta = self._hypervisor_data_by_blocks
+        
+        # get pool data   
+        pool_data = self._pool_data
+
+        # prepare n fill data structure
+        data_by_hypervisor = {} 
+            #  { <hypervisor id> : [ <initial data>, <current data> ]
+            #                        <data> includes "self._get_hypervisor_data_at_block()", "block" and "token0_usd_price" , "token1_usd_price" fields
+                    
+        # only start and end block datas are used
+        for idx, block in enumerate([self.data["initial_block"], self.data["current_block"]]):
+            for hypervisor in hypervisor_dta[block]:
+                
+                pool = None
+                try:
+                    tick_id = self.tick_id(block=block, address=hypervisor["pool"]["id"], baseLower=hypervisor["baseLower"], baseUpper=hypervisor["baseUpper"], limitLower=hypervisor["limitLower"], limitUpper=hypervisor["limitUpper"])
+                    pool = pool_data[tick_id]
+                except:
+                    continue
+
+                # add to structure
+                if not hypervisor["id"] in data_by_hypervisor:
+                    data_by_hypervisor[hypervisor["id"]] = [dict(),dict()] # init block , end block data 
+            
+                # convert types
+                hypervisor = self._convert_dataTypes(hypervisor)
+                
+                # add block and timestamp to hypervisor
+                hypervisor["block"] = block
+                hypervisor["timestamp"] = self._block_ts_map[block]
+
+                # add usd prices
+                hypervisor["token0_usd_price"],hypervisor["token1_usd_price"] = self._calc_USD_prices(hypervisor=hypervisor)
+
+                
+                # get uncollected base fees
+                decimals_0 = int(hypervisor["pool"]["token0"]["decimals"])
+                decimals_1 = int(hypervisor["pool"]["token1"]["decimals"])
+
+                try:
+                    base_fees_0, base_fees_1 = Fees.calc_fees(
```

Comment:
This part looks quite similar to Fees._hypervisor_fees().  Are we able to re-use that?
we can do a larger refactor if required.

Reply:
I think all those need a clearer refactoring of the Data vs Calculations model:

So all data classes could have a "raw_data" and a "processed_data":
"raw_data" being queries to thegraph or other sources with some minor formatting
"processed_data" being the calculations

we could even make those dict so that raw_data["FeesData"] = {}  and raw_data["YieldData"] = {}   and raw_data["ImpermanentData"] = {}  are all in one same class instance ...  ( same with processed)

Also, def formulas going to Calculations and formatting defs going to Data classes
Does this make sense to u?

Reply:
TODO